# Assignment 4 - Classifier

Note: Sorry i missed your class when you were explaining the assignment 4. 
I had to make a few assumptions to complete the work.

## Q1. Use a Neural Network for the implementation of AND and OR functions by using R or Python languages.

### AND Function

Assumption 1: The Neural Network is just one Perceptron where the inputs get multiplied by the weights and sent through a sigmoid activation funtion

In [61]:
import numpy as np

#w1=1, w2=1, b=1.5
def perceptronAND(inputs):
    
    y_temp = inputs[0]*1 + inputs[1]*1 -1.5
    return activation(y_temp) 
     
#w1=1, w2=1, b=0.5
def perceptronOR(inputs):
    
    y_temp = inputs[0]*1 + inputs[1]*1 -0.5
    return activation(y_temp)

def activation(x):
    if x>=0.5:
        return 1
    return 0

print("input AND OR")

for x in [np.array([0,0]), np.array([0,1]), np.array([1,0]), np.array([1,1])]:
    y = perceptronAND(np.array(x))
    z = perceptronOR(np.array(x))
    print(x, y, ' ',z)


input AND OR
[0 0] 0   0
[0 1] 0   1
[1 0] 0   1
[1 1] 1   1


### Q2. Please answer the following questions related to SVM
#### 1. In SVM what is the meaning of margin? Which are the equationsof the two margin Hyperplanes H+ and H-?

A margin is the distance from the decision surface to the closest data point.
SVM's attempt to maximize this margin.

The two hyperplanes H1 and H2 are:

$H1:wx_i+b=+1$

$H2:wx_i+b=-1$

The points on the planes are the tipcs of the support Vectors.

#### 2. Consider the three linearly separable two-dimensional input vectors in the following figure. Find the linear SVM that optimally separates the classes by maximizing the margin.

Assumption 2: That we just point out the function that appears to be allready drawn.

In this situation, all 3 points are support vectors. H+ is the line ontop of the two plus signs and H- is parallel to it and touching the negative sign. The decision boundary is thus halfway between the two margin hyperplanes.
$-x+2=0$


#### 3. What is a kernel function

It is a method of using a linear classifier to solve a non-linear problem. The Kernel function gets applied to the data in order to map it from non-linear into a higher dimensional space where they become separable.

### Q3. Compare Neural Network and SVM in Classification of heart disease data set in R or Python

Assumption 3: That we have to implement it in python


##### Looking back at the Neural Network results:
Accuracy Logistic Regression --> 0.855 
MAE Logistic Regression --> 0.155 
Accuracy Artificial NN --> 0.86
MAE Artificial NN --> 0.144

Lets try to run an SVM on the same data.

In [198]:
#CODE RE-USED FROM PREVIOUS ASSIGNMENT --> SKIP TO 'NEW CODE' LINE =======================================
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn import svm
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score
from sklearn.svm import SVC

url="https://archive.ics.uci.edu/ml/machine-learning-databases/heart-disease/processed.cleveland.data"
missing_values = ["?", "n/a", "--"]
heartData_uncleaned = pd.read_csv(url, na_values=missing_values, header=None)
heartData_uncleaned = pd.DataFrame(heartData_uncleaned)
heartData_uncleaned = heartData_uncleaned.set_axis(["age", "sex", "cp", "trestbps", "chol", "fbs", "restecg", "thalach", "exang", "oldpeak", "slope", "ca", "thal", "num"], axis='columns', inplace=False)
heartData = heartData_uncleaned.dropna()
train, test = train_test_split(heartData, test_size=0.3)
X_train = train.iloc[:,0:13]
y_train = train.iloc[:,-1]
X_test = test.iloc[:,0:13]
y_test = test.iloc[:,-1]
train_binary = train.copy()
test_binary = test.copy()
for index, row in train_binary.iterrows():
    ifor_val = 1;
    if row['num'] != 0:
        train_binary.set_value(index, 'num', ifor_val)
for index, row in test_binary.iterrows():
    ifor_val = 1;
    if row['num'] != 0:
        test_binary.set_value(index, 'num', ifor_val)
X_train_binary = train_binary.iloc[:,0:13]
y_train_binary = train_binary.iloc[:,-1]
X_test_binary = test_binary.iloc[:,0:13]
y_test_binary = test_binary.iloc[:,-1]


#NEW CODE ==================================================================================================
y_train_binary_df = pd.DataFrame(data = y_train_binary).astype(float)
logregBinary = LogisticRegression().fit(X_train_binary, y_train_binary)

clf = svm.SVC()
clf.fit(X_train_binary, y_train_binary_df)

y_pred = clf.predict(X_test_binary)
print(confusion_matrix(y_test_binary, y_pred))
print(classification_report(y_test_binary, y_pred))
print(accuracy_score(y_test_binary, y_pred))

[[52  1]
 [35  2]]
             precision    recall  f1-score   support

          0       0.60      0.98      0.74        53
          1       0.67      0.05      0.10        37

avg / total       0.63      0.60      0.48        90

0.6


c:\python35\lib\site-packages\ipykernel_launcher.py:27: FutureWarning: set_value is deprecated and will be removed in a future release. Please use .at[] or .iat[] accessors instead
c:\python35\lib\site-packages\ipykernel_launcher.py:31: FutureWarning: set_value is deprecated and will be removed in a future release. Please use .at[] or .iat[] accessors instead


## That was linear SVM. It got an accuracy of 0.6
## Lets try non-linear with C=5

In [187]:
clfp = SVC(kernel='poly', degree=8, C=5.0)  
clfp.fit(X_train_binary, y_train_binary) 

SVC(C=5.0, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape='ovr', degree=8, gamma='auto', kernel='poly',
  max_iter=-1, probability=False, random_state=None, shrinking=True,
  tol=0.001, verbose=False)

In [188]:
y_pred = clfp.predict(X_test_binary)

In [189]:
print(confusion_matrix(y_test_binary, y_pred))
print(classification_report(y_test_binary, y_pred))
print(accuracy_score(y_test_binary, y_pred))

[[42 12]
 [ 7 29]]
             precision    recall  f1-score   support

          0       0.86      0.78      0.82        54
          1       0.71      0.81      0.75        36

avg / total       0.80      0.79      0.79        90

0.7888888888888889


We get a very similar accuracy.

Accuracy Artificial NN --> 0.86

SVM = 0.78

I think with some tweaking, the value can go higher. However, the SVM algorithm suffers from the curse of dimensionality.